### PR TITLE
[Feature] Add cached mosaic, cached mixup, and pipeline switch hook.

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add cached mosaic and cached mixup augmentation.

### CODE REVIEW

File: mmdet/datasets/transforms/transforms.py [hunk 1]
```diff
@@ -3005,3 +3005,461 @@ def __repr__(self):
...
+
+        Args:
+            results (dict): Result dict.
+
+        Returns:
+            dict: Updated result dict.
+        """
+        # cache and pop images
+        self.results_cache.append(copy.deepcopy(results))
+        if len(self.results_cache) > self.max_cached_images:
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:3115]: Does `self.max_cached_images` property need to be asserted, if it is less than 4, does it not meet expectations?

File: mmdet/datasets/transforms/transforms.py [hunk 2]
```diff
@@ -3005,3 +3005,461 @@ def __repr__(self):
...
+
+        if len(self.results_cache) <= 4:
+            return results
+
+        if random.uniform(0, 1) > self.prob:
+            return results
+        indices = self.get_indexes(self.results_cache)
+        mix_results = [copy.deepcopy(self.results_cache[i]) for i in indices]
+
+        mosaic_bboxes = []
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:3130]: Is it possible to call super logic?

File: mmdet/datasets/transforms/transforms.py [hunk 3]
```diff
@@ -3005,3 +3005,461 @@ def __repr__(self):
...
+        repr_str += f'center_ratio_range={self.center_ratio_range}, '
+        repr_str += f'pad_val={self.pad_val}, '
+        repr_str += f'prob={self.prob}, '
+        repr_str += f'max_cached_images={self.max_cached_images}, '
+        repr_str += f'random_pop={self.random_pop})'
+        return repr_str
+
+
+@TRANSFORMS.register_module()
+class CachedMixUp(BaseTransform):
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:3220]: Why not inherit mixup?

File: mmdet/datasets/transforms/transforms.py (hunk 2 above)
Comment (RangiLyu) [reply mmdet/datasets/transforms/transforms.py:3130]: Need to refactor Mosaic's logic.

File: mmdet/datasets/transforms/transforms.py (hunk 3 above)
Comment (RangiLyu) [reply mmdet/datasets/transforms/transforms.py:3220]: Can not reuse any method from MixUp.

File: mmdet/datasets/transforms/transforms.py [hunk 4]
```diff
@@ -3005,3 +3005,461 @@ def __repr__(self):
...
+        center_y = int(
+            random.uniform(*self.center_ratio_range) * self.img_scale[0])
+        center_position = (center_x, center_y)
+
+        loc_strs = ('top_left', 'top_right', 'bottom_left', 'bottom_right')
+        for i, loc in enumerate(loc_strs):
+            if loc == 'top_left':
+                results_patch = copy.deepcopy(results)
+            else:
+                results_patch = copy.deepcopy(mix_results[i - 1])
```
Comment (Czm369) [inline mmdet/datasets/transforms/transforms.py:3156]: Add a function `get_mix_results` in `Mosaic`, and rewrite it in `CachedMosaic`.

Comment (RangiLyu) [reply mmdet/datasets/transforms/transforms.py:3156]: Added some TODO

File: mmdet/datasets/transforms/transforms.py [hunk 5]
```diff
@@ -3005,3 +3005,474 @@ def __repr__(self):
...
+        Args:
+            results (dict): Result dict.
+
+        Returns:
+            dict: Updated result dict.
+        """
+        # cache and pop images
+        self.results_cache.append(copy.deepcopy(results))
+        if len(self.results_cache) > self.max_cached_images:
+            if self.random_pop:
```
Comment (Czm369) [inline mmdet/datasets/transforms/transforms.py:3120]: Add a unit test for `results_cache`.

File: mmdet/datasets/transforms/transforms.py [hunk 6]
```diff
@@ -3005,3 +3005,474 @@ def __repr__(self):
...
+        """MixUp transform function.
+
+        Args:
+            results (dict): Result dict.
+
+        Returns:
+            dict: Updated result dict.
+        """
+        # cache and pop images
+        self.results_cache.append(copy.deepcopy(results))
```
Comment (Czm369) [inline mmdet/datasets/transforms/transforms.py:3352]: Add a unit test for `results_cache`.

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**